db/db_test2.cc

```diff
@@ -1268,6 +1268,146 @@ class CompactionCompressionListener : public EventListener {
+      "UncompressBlockContentsForCompressionType:"
+      "TamperWithDecompressionOutput",
+      [&curr_compression_failure_type](void* arg) {
+        if (curr_compression_failure_type == kTestDecompressionCorruption) {
+          BlockContents* contents = static_cast<BlockContents*>(arg);
+          std::string fake_data(contents->data.data(), contents->data.size());
+          // Ensure uncompressed data != original data
+          fake_data.push_back('A');
+          Slice fake_slice = fake_data;
+          *contents = BlockContents(fake_slice);
+        }
+      });
+
+  std::vector<CompressionType> compression_types = GetSupportedCompressions();
+  std::vector<uint32_t> compression_max_dict_bytes = {0, 10};
+  std::vector<uint32_t> compression_parallel_threads = {1, 4};
+
+  std::map<std::string, std::string> key_value_written;
+
+  const int kKeySize = 5;
+  const int kValUnitSize = 16;
+  const int kValSize = 256;
+  Random rnd(405);
+
+  Status s = Status::OK();
+
+  for (auto compression_failure_type : compression_failure_types) {
+    curr_compression_failure_type = compression_failure_type;
+    for (auto compression_type : compression_types) {
+      if (compression_type == kNoCompression) {
+        continue;
+      }
+      for (auto parallel_threads : compression_parallel_threads) {
+        for (auto max_dict_bytes : compression_max_dict_bytes) {
+          options.compression = compression_type;
+          options.compression_opts.parallel_threads = parallel_threads;
+          options.compression_opts.max_dict_bytes = max_dict_bytes;
+          options.bottommost_compression_opts.parallel_threads =
+              parallel_threads;
+          options.bottommost_compression_opts.max_dict_bytes = max_dict_bytes;
+
+          DestroyAndReopen(options);
+          // Write 10 random files
+          for (int i = 0; i < 10; i++) {
+            for (int j = 0; j < 5; j++) {
+              std::string key = RandomString(&rnd, kKeySize);
+              // Ensure good compression ratio
+              std::string valueUnit = RandomString(&rnd, kValUnitSize);
+              std::string value;
+              for (int k = 0; k < kValSize; k += kValUnitSize) {
+                value += valueUnit;
+              }
+              s = Put(key, value);
+              if (compression_failure_type == kTestCompressionFail) {
+                key_value_written[key] = value;
+                ASSERT_OK(s);
+              }
+            }
+            s = Flush();
+            if (compression_failure_type == kTestCompressionFail) {
+              ASSERT_OK(s);
+            }
+            s = dbfull()->TEST_WaitForCompact();
+            if (compression_failure_type == kTestCompressionFail) {
+              ASSERT_OK(s);
+            }
+            if (i == 4) {
+              // Make compression fail at the mid of table building
+              ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->EnableProcessing();
```

Comment:
Move `EnableProcessing` here to simulate failing at the mid of table building. This will expose "passing the same reference" bug and "block_rep->first_key_in_next_block nullptr" bug mentioned in the updated PR description, and they're fixed.